src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -604,7 +635,176 @@ async def create_pex(
         if log_output:
             logger.info("%s", log_output)
 
-    return Pex(digest=result.output_digest, name=request.output_filename, python=python)
+    digest = (
+        await Get(
+            Digest, MergeDigests((result.output_digest, *(pex.digest for pex in request.pex_path)))
+        )
+        if request.pex_path
+        else result.output_digest
+    )
+
+    return BuildPexResult(
+        result=result, pex_filename=request.output_filename, digest=digest, python=python
+    )
+
+
+@rule
+async def create_pex(request: PexRequest) -> Pex:
+    result = await Get(BuildPexResult, PexRequest, request)
+    return result.create_pex()
+
+
+@dataclass(frozen=True)
+class Script:
+    path: PurePath
+
+    @property
+    def argv0(self) -> str:
+        return f"./{self.path}" if self.path.parent == PurePath() else str(self.path)
+
+
+@dataclass(frozen=True)
+class VenvScript:
+    script: Script
+    content: FileContent
+
+
+@dataclass(frozen=True)
+class VenvScriptWriter:
+    pex: Pex
+    venv_dir: PurePath
+
+    def _create_venv_script(
+        self,
+        bash: BashBinary,
+        pex_environment: PexEnvironment,
+        *,
+        script_path: PurePath,
+        venv_executable: PurePath,
+    ) -> VenvScript:
+        env_vars = (
+            f"{name}={shlex.quote(value)}"
+            for name, value in pex_environment.environment_dict(python_configured=True).items()
+        )
+        target_venv_executable = shlex.quote(str(venv_executable))
+        venv_dir = shlex.quote(str(self.venv_dir))
+        execute_pex_args = " ".join(
+            shlex.quote(arg)
+            for arg in pex_environment.create_argv(self.pex.name, python=self.pex.python)
+        )
+
+        script = dedent(
+            f"""\
+            #!{bash.path}
+            set -euo pipefail
+
+            export {" ".join(env_vars)}
+
+            # If the seeded venv has been removed from the PEX_ROOT, we re-seed from the original
+            # `--venv` mode PEX file.
+            if [ ! -e {target_venv_executable} ]; then
+                rm -rf {venv_dir} || true
+                PEX_INTERPRETER=1 {execute_pex_args} -c ''
+            fi
+
+            exec {target_venv_executable} "$@"
+            """
+        )
```

Comment:
Is this logic something that should graduate into PEX, given that it already creates a venv launcher script? The re-creation of the seeded venv seems like something that should be handled there at least.

Reply:
It is in Pex injected in the `__main__.py` in a `--venv` mode PEX file: https://github.com/pantsbuild/pex/pull/1153/files#diff-8018b797310f9896a6574a746376acaf0730b409d6fe03ab73887f823240505cR57

When you create a `--venv` mode PEX you can run the PEX file itself like normal and it will start up using the `__main__.py` and minimal `.bootstrap/` code in the PEX zip file in order to check to see if has already created a venv under `~/.pex/venvs`:
a. If it has, it re-execs itself against the venv.
b. If it hasn't, it seeds the venv and runs from there.

This makes using the PEX file transparent and robust but costs 60ms of overhead over and above a raw venv. I.E.: option a requires ~60ms to check and re-exec.

One way to get around this is to manually create a venv from the PEX once, i.e.: `PEX_TOOLS=1 ./pytest_runner.pex venv pytest.venv` and then store `venv pytest.venv` as an output directory and run that venv for pytest executions. The issue with that is it introduces ~500 ms overhead extra in chroot creation time by the engine / loading time by python / the page cache.

For example:
```
$ hyperfine --warmup 2 './pants --pex-create-explicit-venv test --force src/python/pants/backend/python/util_rules/pex_cli_test.py' './pants test --force src/python/pants/backend/python/util_rules/pex_cli_test.py'
Benchmark #1: ./pants --pex-create-explicit-venv test --force src/python/pants/backend/python/util_rules/pex_cli_test.py
  Time (mean ± σ):      2.412 s ±  0.065 s    [User: 524.8 ms, System: 55.3 ms]
  Range (min … max):    2.285 s …  2.503 s    10 runs
 
Benchmark #2: ./pants test --force src/python/pants/backend/python/util_rules/pex_cli_test.py
  Time (mean ± σ):      1.878 s ±  0.070 s    [User: 621.3 ms, System: 79.4 ms]
  Range (min … max):    1.753 s …  1.964 s    10 runs
 
Summary
  './pants test --force src/python/pants/backend/python/util_rules/pex_cli_test.py' ran
    1.28 ± 0.06 times faster than './pants --pex-create-explicit-venv test --force src/python/pants/backend/python/util_rules/pex_cli_test.py'
```

That's with some changes over in PEX to support creating venvs that use no symlinks and precompile all .py files in the entire venv:
```diff
diff --git a/src/python/pants/backend/python/util_rules/pex.py b/src/python/pants/backend/python/util_rules/pex.py
index 0ea56f37a..07401ec2b 100644
--- a/src/python/pants/backend/python/util_rules/pex.py
+++ b/src/python/pants/backend/python/util_rules/pex.py
@@ -767,11 +767,164 @@ def wrap_venv_prex_request(pex_request: PexRequest) -> VenvPexRequest:
     return VenvPexRequest(pex_request)
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class PexProcess:
+    pex: Pex
+    argv: Tuple[str, ...]
+    description: str = dataclasses.field(compare=False)
+    level: LogLevel
+    input_digest: Optional[Digest]
+    extra_env: Optional[FrozenDict[str, str]]
+    output_files: Optional[Tuple[str, ...]]
+    output_directories: Optional[Tuple[str, ...]]
+    timeout_seconds: Optional[int]
+    execution_slot_variable: Optional[str]
+    cache_scope: Optional[ProcessCacheScope]
+
+    def __init__(
+        self,
+        pex: Pex,
+        *,
+        description: str,
+        argv: Iterable[str] = (),
+        level: LogLevel = LogLevel.INFO,
+        input_digest: Optional[Digest] = None,
+        extra_env: Optional[Mapping[str, str]] = None,
+        output_files: Optional[Iterable[str]] = None,
+        output_directories: Optional[Iterable[str]] = None,
+        timeout_seconds: Optional[int] = None,
+        execution_slot_variable: Optional[str] = None,
+        cache_scope: Optional[ProcessCacheScope] = None,
+    ) -> None:
+        self.pex = pex
+        self.argv = tuple(argv)
+        self.description = description
+        self.level = level
+        self.input_digest = input_digest
+        self.extra_env = FrozenDict(extra_env) if extra_env else None
+        self.output_files = tuple(output_files) if output_files else None
+        self.output_directories = tuple(output_directories) if output_directories else None
+        self.timeout_seconds = timeout_seconds
+        self.execution_slot_variable = execution_slot_variable
+        self.cache_scope = cache_scope
+
+
+@dataclass(frozen=True)
+class CopiedVenvScriptWriter:
+    pex: Pex
+    venv_dir: PurePath
+
+    def _create_venv_script(
+        self,
+        bash: BashBinary,
+        pex_environment: PexEnvironment,
+        *,
+        script_path: PurePath,
+        venv_executable: Optional[PurePath] = None,
+    ) -> VenvScript:
+        env_vars = (
+            f"{name}={shlex.quote(value)}"
+            for name, value in pex_environment.environment_dict(python_configured=True).items()
+        )
+        target_venv_executable = shlex.quote(str(venv_executable)) if venv_executable else ""
+        script = dedent(
+            f"""\
+            #!{bash.path}
+            set -euo pipefail
+
+            export {" ".join(env_vars)}
+
+            HERE="$(cd "$(dirname "${{BASH_SOURCE[0]}}")" && pwd)"
+
+            exec ${{HERE}}/{self.venv_dir}/bin/python {target_venv_executable} "$@"
+            """
+        )
+        return VenvScript(
+            script=Script(script_path),
+            content=FileContent(path=str(script_path), content=script.encode(), is_executable=True),
+        )
+
+    def exe(self, bash: BashBinary, pex_environment: PexEnvironment) -> VenvScript:
+        """Writes a safe shim for the venv's executable `pex` script."""
+        script_path = PurePath(f"{self.pex.name}_pex_shim.sh")
+        return self._create_venv_script(
+            bash, pex_environment, script_path=script_path, venv_executable=self.venv_dir / "pex"
+        )
+
+    def bin(self, bash: BashBinary, pex_environment: PexEnvironment, name: str) -> VenvScript:
+        """Writes a safe shim for an executable or script in the venv's `bin` directory."""
+        script_path = PurePath(f"{self.pex.name}_bin_{name}_shim.sh")
+        return self._create_venv_script(
+            bash,
+            pex_environment,
+            script_path=script_path,
+            venv_executable=self.venv_dir / "bin" / name,
+        )
+
+    def python(self, bash: BashBinary, pex_environment: PexEnvironment) -> VenvScript:
+        """Writes a safe shim for the venv's python binary."""
+        script_path = PurePath(f"{self.pex.name}_bin_python_shim.sh")
+        return self._create_venv_script(bash, pex_environment, script_path=script_path)
+
+
 @rule
 async def create_venv_pex(
-    request: VenvPexRequest, bash: BashBinary, pex_environment: PexEnvironment
+    request: VenvPexRequest,
+    bash: BashBinary,
+    pex_environment: PexEnvironment,
+    pex_runtime_environment: PexRuntimeEnvironment,
 ) -> VenvPex:
     pex_request = request.pex_request
+    if pex_runtime_environment.create_explicit_venv:
+        tools_pex_request = dataclasses.replace(
+            pex_request, additional_args=pex_request.additional_args + ("--venv", "--include-tools")
+        )
+        tools_pex_result = await Get(BuildPexResult, PexRequest, tools_pex_request)
+        tools_pex = tools_pex_result.create_pex()
+
+        venv_dir = PurePath(f"{pex_request.output_filename}.venv")
+        create_venv_result = await Get(
+            ProcessResult,
+            PexProcess(
+                tools_pex,
+                argv=["venv", "--collisions-ok", "--copy", "--compile", str(venv_dir)],
+                extra_env={"PEX_TOOLS": "1"},
+                output_directories=[str(venv_dir)],
+                description=f"Create a venv for {request.pex_request.output_filename}.",
+            ),
+        )
+
+        copied_venv_script_writer = CopiedVenvScriptWriter(pex=tools_pex, venv_dir=venv_dir)
+        pex = copied_venv_script_writer.exe(bash, pex_environment)
+        python = copied_venv_script_writer.python(bash, pex_environment)
+        scripts = {
+            bin_name: copied_venv_script_writer.bin(bash, pex_environment, bin_name)
+            for bin_name in request.bin_names
+        }
+        scripts_digest = await Get(
+            Digest,
+            CreateDigest(
+                (
+                    pex.content,
+                    python.content,
+                    *(venv_script.content for venv_script in scripts.values()),
+                )
+            ),
+        )
+        input_digest = await Get(
+            Digest, MergeDigests((create_venv_result.output_digest, scripts_digest))
+        )
+
+        return VenvPex(
+            digest=input_digest,
+            pex=pex.script,
+            python=python.script,
+            bin=FrozenDict(
+                (bin_name, venv_script.script) for bin_name, venv_script in scripts.items()
+            ),
+        )
+
     seeded_venv_request = dataclasses.replace(
         pex_request, additional_args=pex_request.additional_args + ("--venv", "--seed")
     )
@@ -850,49 +1003,6 @@ async def two_step_create_pex(two_step_pex_request: TwoStepPexRequest) -> TwoSte
     return TwoStepPex(pex=full_pex)
 
 
-@frozen_after_init
-@dataclass(unsafe_hash=True)
-class PexProcess:
-    pex: Pex
-    argv: Tuple[str, ...]
-    description: str = dataclasses.field(compare=False)
-    level: LogLevel
-    input_digest: Optional[Digest]
-    extra_env: Optional[FrozenDict[str, str]]
-    output_files: Optional[Tuple[str, ...]]
-    output_directories: Optional[Tuple[str, ...]]
-    timeout_seconds: Optional[int]
-    execution_slot_variable: Optional[str]
-    cache_scope: Optional[ProcessCacheScope]
-
-    def __init__(
-        self,
-        pex: Pex,
-        *,
-        description: str,
-        argv: Iterable[str] = (),
-        level: LogLevel = LogLevel.INFO,
-        input_digest: Optional[Digest] = None,
-        extra_env: Optional[Mapping[str, str]] = None,
-        output_files: Optional[Iterable[str]] = None,
-        output_directories: Optional[Iterable[str]] = None,
-        timeout_seconds: Optional[int] = None,
-        execution_slot_variable: Optional[str] = None,
-        cache_scope: Optional[ProcessCacheScope] = None,
-    ) -> None:
-        self.pex = pex
-        self.argv = tuple(argv)
-        self.description = description
-        self.level = level
-        self.input_digest = input_digest
-        self.extra_env = FrozenDict(extra_env) if extra_env else None
-        self.output_files = tuple(output_files) if output_files else None
-        self.output_directories = tuple(output_directories) if output_directories else None
-        self.timeout_seconds = timeout_seconds
-        self.execution_slot_variable = execution_slot_variable
-        self.cache_scope = cache_scope
-
-
 @rule
 async def setup_pex_process(request: PexProcess, pex_environment: PexEnvironment) -> Process:
     pex = request.pex
diff --git a/src/python/pants/backend/python/util_rules/pex_cli.py b/src/python/pants/backend/python/util_rules/pex_cli.py
index 19c7bc1b6..793c5f5a2 100644
--- a/src/python/pants/backend/python/util_rules/pex_cli.py
+++ b/src/python/pants/backend/python/util_rules/pex_cli.py
@@ -38,7 +38,7 @@ class PexBinary(TemplatedExternalTool):
     help = "The PEX (Python EXecutable) tool (https://github.com/pantsbuild/pex)."
 
     default_version = "v2.1.30"
-    default_url_template = "https://github.com/pantsbuild/pex/releases/download/{version}/pex"
+    default_url_template = "file:/home/jsirois/dev/pantsbuild/jsirois-pex/dist/pex"
 
     @classproperty
     def default_known_versions(cls):
@@ -47,8 +47,8 @@ class PexBinary(TemplatedExternalTool):
                 (
                     cls.default_version,
                     plat,
-                    "1a25f75ea95cdf78cd0cbb249c0e67de5259604d1a1336b49863dc1a39e0993d",
-                    "2968698",
+                    "45d266ac411825e632c97759dcefa296151dc976ee3d460009a237c072c8f519",
+                    "2969074",
                 )
             )
             for plat in ["darwin", "linux"]
diff --git a/src/python/pants/backend/python/util_rules/pex_environment.py b/src/python/pants/backend/python/util_rules/pex_environment.py
index 944681f17..46e896983 100644
--- a/src/python/pants/backend/python/util_rules/pex_environment.py
+++ b/src/python/pants/backend/python/util_rules/pex_environment.py
@@ -66,6 +66,13 @@ class PexRuntimeEnvironment(Subsystem):
                 "Set the verbosity level of PEX logging, from 0 (no logging) up to 9 (max logging)."
             ),
         )
+        register(
+            "--create-explicit-venv",
+            advanced=True,
+            type=bool,
+            default=False,
+            help=("Whether venv PEXes for internal use should create explicit venvs to run from."),
+        )
 
     @memoized_property
     def path(self) -> Tuple[str, ...]:
@@ -92,6 +99,10 @@ class PexRuntimeEnvironment(Subsystem):
             raise ValueError("verbosity level must be between 0 and 9")
         return level
 
+    @property
+    def create_explicit_venv(self) -> bool:
+        return cast(bool, self.options.create_explicit_venv)
+
 
 class PythonExecutable(BinaryPath, EngineAwareReturnType):
     """The BinaryPath of a Python executable."""

```


The other way to get around it is to use `--seed` mode where PEX prints out the path to the seeded venv. That allows you to execute the seeded venv directly, skipping the PEX file overhead, but it open you up to the world changing when someone nukes the `~/.pex/venvs` cache out from under you.

As such, for now, the only option to get raw venv speed is `--venv` `--seed` and a wrapper script to repair the `--seeded` venv when it disappears. The wrapping bash script only costs ~1ms.

Reply:
Thats _alot_ of text, but I did not want to shirk my duty to explain everything going on. The elevator is important to keep in mind as we move forward with pants and new languages etc:
1. materializing items in the chroot is slow - we get away with not having to internalize this lesson only because we have named caches for the Python stack and these allow us to store single zip files that are only exploded once and centrally to the PEX_ROOT.
2. thrashing the page cache may also be slow - when we store lots of files in LMDB and then materialize in a chroot, they are all as if new files to the OS and must be read in cold.

I'm not sure how much of the 500+ms overhead is due to 1 vs 2, but its simply super expensive to store big dirs in LMDB and then try to use them later to run fine grained things.

This would be a fun problem to work on and we have some ideas including brfs, but for now we can leverage named caches.

Reply:
I did come up with one way for this logic to graduate to Pex, but Pex would have to grow a dependency on bash. Basically you'd need to replace the current python shebang with the option to use a `#!/usr/bin/env bash`. In addition to the logic the current bash script has, the new bash script would also need to learn how to find an interpreter useable to bootstrap any pex (we have that logic in rules currently). That's all doable but not super appealing to support what appears to be a Panst only use case. Any other use case would reach for a one-time `PEX_TOOLS=1 ./my.pex venv here/please` and then run the app from that `here/please/pex` forevermore.

Reply:
> Thats alot of text, but I did not want to shirk my duty to explain everything going on. 

Thanks a lot, this helps. It would be good to (if you haven't already? sorry, haven't re-reviewed) inline some of this as a comment.

If I understood this, this script ends up avoiding 60ms overhead of using the `--venv` mode's detection of the existing extracted venv? Honestly, it might be that it would be worth it to have slightly simpler code (and fewer comments explaining it) and to just eat the 60ms for now (with a much smaller comment explaining that we have 60ms there waiting for us if we want to get it in the future)?

The move to `--venv` mode seems to remove pytest as the long pole in our runs, with/without the 60ms. But up to you!

Reply:
> The elevator is important to keep in mind as we move forward with pants and new languages etc:
> [..]
> I'm not sure how much of the 500+ms overhead is due to 1 vs 2, but its simply super expensive to store big dirs in LMDB and then try to use them later to run fine grained things.

This would be super useful to quantify, I think. I had meant to add workunits wrapping the "sandbox creation" vs "process execution" portions of runs to gather more data here, but haven't yet.

Reply:
> If I understood this, this script ends up avoiding 60ms overhead of using the --venv mode's detection of the existing extracted venv?

Yes.

> Honestly, it might be that it would be worth it to have slightly simpler code (and fewer comments explaining it) and to just eat the 60ms for now (with a much smaller comment explaining that we have 60ms there waiting for us if we want to get it in the future)?
> The move to --venv mode seems to remove pytest as the long pole in our runs, with/without the 60ms. But up to you!

Yeah, I'm going to stick with the 60ms win. This was a good bit of the work I put in back in December over in Pex with --seed and various tests. Back then as now it seems to me Pants is my enemy if it makes what I already do significantly slower. Adding 60ms to a 200ms test is pretty bad. It just happens to be dwarfed by Pants startup overhead today. I'd definitely rather have the python ruleset out of the way for when that issue is fixed.

I'll circle back with a PR to add comments in the right spots about what's going on here in the large.